### DISCUSSION TITLE
Steps toward Frigate HA

### DISCUSSION THREAD

**Discussion by quenthal** [blakeblackshear/frigate#5309] [Show and tell]
I've planned to enable high availability to some of my proxmox containers and virtual machines - including Frigate. I'm in no way script wizard, and there is probably lots of better ways to achieve this, but this is how I solved this by figuring it out, partially with help of ChatGPT.

Everything probably would have been easy if all hosts/ nodes were equal regarding detecters. However, each server on my environment has different detectors. As (at least with my setup) Frigate will eventually crash if it is configured to multiple detectors while some of those detectors are missing, I decided I start with replacing config.yml automatically on startup depending on which version of Coral my host has. I have Coral USB on host 1 and Coral PCI on host 2. This can easily be adjusted to CPU/nVidia detectors as well.

1. I started with normal setup on hosts and with LXC configs. 

2. Make sure that Coral PCI-drivers are installed. Make sure that both USB and PCIE Corals are passed thru to Frigate LXC even though only one of them is really installed on each server. This is to make sure that LXC config is identical even if differect Coral is used on different hosts, ie. I have both of these lines in the lxc config (apex_0 for PCIE Coral, and passing thru usb/002 as a whole in my case for USB Coral):
```
lxc.mount.entry: /dev/apex_0 dev/apex_0 none bind,optional,create=file
lxc.mount.entry: /dev/bus/usb/002/ dev/bus/usb/002/ none bind,optional,create=dir
```

3. Naturally docker is identical as well, as it runs within LXC, ie. it has both usb and PCIE passed thru from LXC:
`... --device /dev/bus/usb:/dev/bus/usb --device /dev/apex_0 --device /dev/dri/renderD128 -v /config/config.yml:/config/config.yml:ro ...`

4. As can be seen on step 3, I'm passing thru /config/config.yml from LXC host to docker. I've infact 3 config yamls, with single difference; the detector part. With /config/config_pci.yml:
```
...
detectors:
  coral_pci:
    type: edgetpu
    device: pci
```
..and with /config/config_usb.yml:
```
...
detectors:
  coral_usb:
    type: edgetpu
    device: usb
```
Otherwise the configs are similar.

5. I then added the following script as a service within LXC to make adjustments for the Frigate docker config depending on which detector is installed on the host. 
So i begun by creating the script itself by `nano /config/checkDetector.sh`:

```
#!/bin/bash

if lsusb | grep 18d1:9302; then
  echo "Coral USB found"
  cp /config/config_usb.yml /config/config.yml
else
  echo "Coral USB not found"
  cp /config/config_pci.yml /config/config.yml
fi
```
Then I made it runnable by `chmod +x /config/checkDetector.sh`

After that I made it run as a service before docker `nano /etc/systemd/system/checkFrigateDetector.service`:
```
[Unit]
Description=Check which detector is to be used with Frigate
Before=docker.service

[Service]
Type=oneshot
ExecStart=/config/checkDetector.sh

[Install]
WantedBy=multi-user.target docker.service
```
Then I enabled it:
```
systemctl daemon-reload
systemctl enable checkFrigateDetector.service
```

After all this, whenever LXC container is (re)started, it checks if USB Coral is installed before docker is started. If it is, config using it will be replacing the config that frigate will use. If it is not, it is defaulted to use config using pci-e/m.2 version of Coral and that will replace the config Frigate will use after start.

**Comment by bung69** [comment]
are you still using this method without issue?

do you also use storage replication?
